src/PowerShell/Microsoft.WinGet.DSC/Microsoft.WinGet.DSC.psm1

```diff
@@ -0,0 +1,296 @@
+    [DscProperty(Mandatory)]
+    [Hashtable[]]$Sources
+
+    [DscProperty()]
+    [Ensure]$Ensure = [Ensure]::Present
+
+    [DscProperty()]
+    [bool]$Reset = $false
+
+    [DscProperty()]
+    [WinGetAction]$Action = [WinGetAction]::Full
+
+    # Gets the current sources on winget.
+    [WinGetSourcesResource] Get()
+    {
+        Assert-WinGetCommand "Get-WinGetSource"
+        $packageCatalogReferences = Get-WinGetSource
+        $wingetSources = [List[Hashtable]]::new()
+        foreach ($packageCatalogReference in $packageCatalogReferences)
+        {
+            $source = @{
+                Arg = $packageCatalogReference.Info.Argument
+                Identifier = $packageCatalogReference.Info.Id
+                Name = $packageCatalogReference.Info.Name
+                Type = $packageCatalogReference.Info.Type
+            }
+            $wingetSources.Add($source)
+        }
+
+        $result = @{
+            SID = ''
+            Sources = $wingetSources
+        }
+        return $result
+    }
+
+    # Tests if desired properties match.
+    [bool] Test()
+    {
+        $currentSources = $this.Get().Sources
+
+        # If this is a full match and the counts are different give up.
+        if (($this.Action -eq [WinGetAction]::Full) -and ($this.Sources.Count -ne $currentSources.Count))
+        {
+            return $false
+        }
+
+        # There's no need to differentiate between Partial and Full anymore.
+        foreach ($source in $this.Sources)
+        {
+            # Require Name and Arg.
+            if ((-not $source.ContainsKey("Name")) -or [string]::IsNullOrWhiteSpace($source.Name))
+            {
+                throw "Invalid source input. Name is required."
+            }
+
+            if ((-not $source.ContainsKey("Arg")) -or [string]::IsNullOrWhiteSpace($source.Arg))
+            {
+                throw "Invalid source input. Arg is required."
+            }
+
+            # Type has a default value.
+            $sourceType = "Microsoft.PreIndexed.Package"
+            if ($source.ContainsKey("Type") -and (-not([string]::IsNullOrWhiteSpace($source.Type))))
+            {
+                $sourceType = $source.Type
+            }
+
+            $result = $currentSources | Where-Object { $_.Name -eq $source.Name -and $_.Arg -eq $source.Arg -and $_.Type -eq $sourceType }
+
+            # Source not found.
+            if ($null -eq $result)
+            {
+                return $false
+            }
+        }
+
+        return $true
+    }
+
+    # Sets the desired properties.
+    [void] Set()
+    {
+        Assert-IsAdministrator
+        Assert-WinGetCommand "Add-WinGetSource"
+        Assert-WinGetCommand "Reset-WinGetSource"
+        Assert-WinGetCommand "Remove-WinGetSource"
+
+        foreach ($source in $this.Sources)
+        {
+            # The call to test already validated that Name and Arg are set and valid.

```

Comment:
I don't think that there is any requirement to call `Test` first, and you have not done so here.  It should be validated in `Set` as well.